internal-services/catalog/pulp-push-disk-images-task.yaml

```diff
@@ -0,0 +1,191 @@
+
+        set +x
+        echo "$exodusGwCert" > "$EXODUS_GW_CERT"
+        echo "$exodusGwKey" > "$EXODUS_GW_KEY"
+        echo "$pulpCert" > "$PULP_CERT"
+        echo "$pulpKey" > "$PULP_KEY"
+        echo "$udcacheCert" > "$UDCACHE_CERT"
+        echo "$udcacheKey" > "$UDCACHE_KEY"
+        set -x
+
+        DISK_IMAGE_DIR="$(mktemp -d)"
+
+        process_component() { # Expected argument is [component json]
+            COMPONENT=$1
+            PULLSPEC=$(jq -r '.containerImage' <<< "${COMPONENT}")
+            DESTINATION="${DISK_IMAGE_DIR}/$(jq -r '.staged.destination' <<< "${COMPONENT}")/FILES"
+            mkdir -p "${DESTINATION}"
+            DOWNLOAD_DIR=$(mktemp -d)
+            cd $DOWNLOAD_DIR
+            # oras has very limited support for selecting the right auth entry,
+            # so create a custom auth file with just one entry
+            AUTH_FILE=$(mktemp)
+            select-oci-auth "${PULLSPEC}" > "$AUTH_FILE"
+            oras pull --registry-config "$AUTH_FILE" "$PULLSPEC"
+            NUM_MAPPED_FILES=$(jq '.staged.files | length' <<< ${COMPONENT})
+            for ((i = 0; i < $NUM_MAPPED_FILES; i++)) ; do
+                FILE=$(jq -c --arg i "$i" '.staged.files[$i|tonumber]' <<< $COMPONENT)
+                SOURCE=$(jq -r '.source' <<< $FILE)
+                FILENAME=$(jq -r '.filename' <<< $FILE)
+                gzip -d "${SOURCE}.gz" || echo "didn't find mapped file: ${SOURCE}.gz"
+                DESTINATION_FILE="${DESTINATION}/${FILENAME}"
+                # Albeit a very small one, this is a race condition since this is run in parallel
+                if [ -f "${DESTINATION_FILE}" ] ; then
+                    echo -n "Multiple files use the same destination value: $DESTINATION"
+                    echo " and filename value: $FILENAME. Failing..."
+                    exit 1
+                fi
+                mv $SOURCE "${DESTINATION_FILE}"
+            done
+        }
+
+        N=$(params.concurrentLimit)  # The maximum number of images to be pulled at once
+        RUNNING_JOBS="\j" # Bash parameter for number of jobs currently running
+        NUM_COMPONENTS=$(jq '.components | length' <<< "$SNAPSHOT_JSON")
+        # Process each component in parallel
+        for ((i = 0; i < $NUM_COMPONENTS; i++)) ; do
+            COMPONENT=$(jq -c --arg i "$i" '.components[$i|tonumber]' <<< "$SNAPSHOT_JSON")
+            # Limit batch size to concurrent limit
+            while (( ${RUNNING_JOBS@P} >= N )); do
+                wait -n
+            done
+            process_component $COMPONENT &
+        done
+
+        # Wait for remaining processes to finish
+        while (( ${RUNNING_JOBS@P} > 0 )); do
+            wait -n
+        done
+
+        # Change to the subdir with the images
+        cd ${DISK_IMAGE_DIR}
+
+        STAGED_JSON='{"header":{"version": "0.2"},"payload":{"files":[]}}'
+
+        # Add the files to the payload
+        for file in $(find * -type f); do
```

Comment:
Are we sure there are no spaces in the paths anywhere? I guess probably there shouldn't be?

Reply:
Yeah, `$SOURCE` should be `"$SOURCE"` - I'm pretty sure the shell check would tell you that once enabled :)

Reply:
I added quotes to `$SOURCE`. Is there something else to change here?

Reply:
It depends. I was trying to point out that this is fragile because of possible spaces and you essentially said that it would break even before this. That is fixed, but that doesn't change the fragility of this. And now I can see that shellcheck doesn't like it either.

I would be fine with it if you're confident there won't be spaces. But now that we have shellcheck, you should probably change it, or you'll have to find a way to tell shellcheck to ignore it.

But I guess you should probably use a while loop as shellcheck suggests (including `find ... -print0`).

Reply:
fixed